### PR TITLE
[Nataliya Mykytyuk] Image zoom

### PR DESCRIPTION
**Pull request by dnistrova**

### CODE REVIEW

File: src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import { Component, OnInit, Input, OnChanges } from '@angular/core';
+
+@Component({
+  selector: 'app-zoomit',
+  templateUrl: './zoomit.component.html',
+  styleUrls: ['./zoomit.component.scss']
+})
+export class ZoomitComponent implements OnInit, OnChanges {
+  // @Input () zsrc: string;
+  zsrc = 'http://www.therxreview.com/wp-content/uploads/2013/05/Reebok-Track-Jacket.png';
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts]: this url should come as Input

File: src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  public zUrl: string;
+  isHovered = false;
+
+  ngOnInit() {
+  this.zUrl = `url(${this.zsrc})`;
+  this.pos = '300% 300%';
+  }
+
+  ngOnChanges() {
+    this.zUrl = `url(${this.zsrc})`;
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts]: why do you need additional zUrl property?

zsc is the same

File: src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+})
+export class ZoomitComponent implements OnInit, OnChanges {
+  // @Input () zsrc: string;
+  zsrc = 'http://www.therxreview.com/wp-content/uploads/2013/05/Reebok-Track-Jacket.png';
+  public pos: string;
+  public zUrl: string;
+  isHovered = false;
+
+  ngOnInit() {
+  this.zUrl = `url(${this.zsrc})`;
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts]: why do you need additional zUrl property?

zsc is the same

File: src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  this.pos = '300% 300%';
+  }
+
+  ngOnChanges() {
+    this.zUrl = `url(${this.zsrc})`;
+  }
+
+  onMouseMove(event: any) {
+  const element = event.target;
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts]: missed tab

File: src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+export class ZoomitComponent implements OnInit, OnChanges {
+  // @Input () zsrc: string;
+  zsrc = 'http://www.therxreview.com/wp-content/uploads/2013/05/Reebok-Track-Jacket.png';
+  public pos: string;
+  public zUrl: string;
+  isHovered = false;
+
+  ngOnInit() {
+  this.zUrl = `url(${this.zsrc})`;
+  this.pos = '300% 300%';
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts]: the default values should be defined above. (ngOnInit should be used for ajax calls, running some logic on start etc, subscription on event etc...)

File: src/app/shared/components/product-item/product-item.component.spec.ts [hunk 6]
```diff
@@ -25,4 +62,71 @@ describe('ProductItemComponent', () => {
...
+
+  it('should return undefined if we call stop', () => {
+    fixture = TestBed.createComponent(ProductItemComponent);
+    component = fixture.componentInstance;
+    component.product = product;
+    const functionsReturn = component.onMouseOver();
+    expect(functionsReturn).toBeUndefined();
+  });
+
+  it('should return undefined if we call stop', () => {
```
Comment (dverbovyi) [inline src/app/shared/components/product-item/product-item.component.spec.ts]: 3 test cases with the same description. Compose into 1

File: src/app/shared/components/product-item/product-item.component.ts [hunk 7]
```diff
@@ -85,7 +80,12 @@ export class ProductItemComponent implements OnInit {
     this.addItemToWishList.emit(this.product);
   }
 
-  private resetDefaultThumbnail(): void {
+  resetDefaultThumbnail(): void {
```
Comment (dverbovyi) [inline src/app/shared/components/product-item/product-item.component.ts]: this method is used only internally. Why did you remove private modification?

File: src/app/product-details-page/components/product-description/product-description.component.html [hunk 8]
```diff
@@ -1,6 +1,9 @@
   <div class="main-container">
     <div class="main-container-wrapper">
-      <div class="image-container"></div>
+      <div class="image-container">
+        <!-- <app-zoomit [zsrc]="imageSrc"></app-zoomit> -->
+        <app-zoomit></app-zoomit>
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-description/product-description.component.html]: the src should be passed as an attribute

File: src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.spec.ts [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+  it('should return undefined if we call stop', () => {
+    fixture = TestBed.createComponent(ZoomitComponent);
+    component = fixture.componentInstance;
+
+    const functionsReturn = component.onMouseLeave();
+    expect(functionsReturn).toBeUndefined();
+  });
+
+  it('should return undefined if we call stop', () => {
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.spec.ts]: 2 test-cases with the same name. Group into 1

**Review by dverbovyi — CHANGES_REQUESTED** [review]
@dnistrova not bad. Please fix formatting and other minor changes

File: src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts (hunk 1 above)
Comment (dnistrova) [reply src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts]: this line will be deleting after integration and this url will come as Input (it's commented)

File: src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts (hunk 2 above)
Comment (dnistrova) [reply src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts]: it is not the same. zsrc is for image source, and zUrl is for background-image (url("zsrc"))

File: src/app/product-details-page/components/product-description/product-description.component.html (hunk 8 above)
Comment (dnistrova) [reply src/app/product-details-page/components/product-description/product-description.component.html]: I will do it after merging with Artur's branch

File: src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts (hunk 3 above)
Comment (dnistrova) [reply src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts]: Look at previous comment.
Also it will changе after clicking on different product's preview images

File: src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts (hunk 4 above)
Comment (dnistrova) [reply src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts]: OK, I'll fix this

File: src/app/shared/components/product-item/product-item.component.spec.ts (hunk 6 above)
Comment (dnistrova) [reply src/app/shared/components/product-item/product-item.component.spec.ts]: OK

File: src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts (hunk 1 above)
Comment (dverbovyi) [reply src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts]: I can't get the point why did you do so. The integration part should be only cover the case of passing the appropriate data from the parent component, and your component has not to be changed.
Just create mock-images in the parent component and pass the src though the input-attribute. After integration, the mock-part will be removed, and replaced, but your component will not be changed

File: src/app/product-details-page/components/product-description/product-description.component.html (hunk 8 above)
Comment (dverbovyi) [reply src/app/product-details-page/components/product-description/product-description.component.html]: please, do it now, and create mocked src in the parent class (and remove from your component)

you should write a code which closes for changes. The integration part has to be done through the described interfaces (input-properties) passed the data from the top-level component to bottom. And your component should not know about the specificity of data (is it mock or real). It should be handled in the same way, without additional changes required

File: src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
+
+  it('should return undefined if we call stop', () => {
+    fixture = TestBed.createComponent(ZoomitComponent);
+    component = fixture.componentInstance;
+
+    const functionsReturn = component.onMouseLeave();
+    expect(functionsReturn).toBeUndefined();
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.spec.ts]: this expect validate nothing

check here onMouseLeave is really called on mouseLeave event

File: src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.spec.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+  });
+
+  it('should return undefined if we call stop', () => {
+    fixture = TestBed.createComponent(ZoomitComponent);
+    component = fixture.componentInstance;
+    const functionsReturn = component.onMouseOver();
+    expect(functionsReturn).toBeUndefined();
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.spec.ts]: this expect validate nothing

check here onMouseOver is really called on mouseOver event

File: src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts (hunk 1 above)
Comment (dverbovyi) [reply src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts]: the mock data should be in parent component

File: src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts (hunk 3 above)
Comment (dverbovyi) [reply src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts]: wrapping in the url should be in the template

try `[style.background-image.url]="zUrl"`

File: src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts (hunk 5 above)
Comment (dverbovyi) [reply src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts]: move the default position to css

File: src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts (hunk 2 above)
Comment (dverbovyi) [reply src/app/product-details-page/components/product-description/product-image-preview/zoomit/zoomit.component.ts]: wrapping in the url should be in the template

try [style.background-image.url]="zUrl"

File: src/app/product-details-page/product-details-page.module.ts [hunk 12]
```diff
@@ -4,9 +4,10 @@ import { CommonModule } from '@angular/common';
 import { ProductDetailsPageRoutingModule } from './product-details-page-routing.module';
 import { ProductDescriptionComponent } from './components/product-description/product-description.component';
 import { ProductDetailsPageComponent } from './components/product-details-page/product-details-page.component';
+import { ZoomitComponent } from './components/product-description/product-image-preview/zoomit/zoomit.component';
```
Comment (dverbovyi) [inline src/app/product-details-page/product-details-page.module.ts]: what's the weird name for the component, zoomit? rename to Zoom or ZoomImage Component

File: src/app/shared/components/product-item/product-item.component.spec.ts [hunk 13]
```diff
@@ -25,4 +62,71 @@ describe('ProductItemComponent', () => {
...
+      `url(${product.thumbnailImageSrc})`
+    );
+  });
+
+  it('should return undefined if we call stop', () => {
+    fixture = TestBed.createComponent(ProductItemComponent);
+    component = fixture.componentInstance;
+    component.product = product;
+    const functionsReturn = component.hoverStateIn();
+    expect(functionsReturn).toBeUndefined();
```
Comment (dverbovyi) [inline src/app/shared/components/product-item/product-item.component.spec.ts]: this test doesn't make sense

File: src/app/shared/components/product-item/product-item.component.spec.ts [hunk 14]
```diff
@@ -25,4 +62,71 @@ describe('ProductItemComponent', () => {
...
+  });
+
+  it('should return undefined if we call stop', () => {
+    fixture = TestBed.createComponent(ProductItemComponent);
+    component = fixture.componentInstance;
+    component.product = product;
+    const functionsReturn = component.hoverStateOut();
+    expect(functionsReturn).toBeUndefined();
```
Comment (dverbovyi) [inline src/app/shared/components/product-item/product-item.component.spec.ts]: this test doesn't make sense

File: src/app/shared/components/product-item/product-item.component.spec.ts [hunk 15]
```diff
@@ -25,4 +62,71 @@ describe('ProductItemComponent', () => {
...
+  });
+
+  it('should return undefined if we call stop', () => {
+    fixture = TestBed.createComponent(ProductItemComponent);
+    component = fixture.componentInstance;
+    component.product = product;
+    const functionsReturn = component.onMouseLeave();
+    expect(functionsReturn).toBeUndefined();
```
Comment (dverbovyi) [inline src/app/shared/components/product-item/product-item.component.spec.ts]: this test doesn't make sense
check the onMouseLeave is really called on mouse-leave event

File: src/app/shared/components/product-item/product-item.component.spec.ts [hunk 16]
```diff
@@ -25,4 +62,71 @@ describe('ProductItemComponent', () => {
...
+  });
+
+  it('should return undefined if we call stop', () => {
+    fixture = TestBed.createComponent(ProductItemComponent);
+    component = fixture.componentInstance;
+    component.product = product;
+    const functionsReturn = component.onMouseOver();
+    expect(functionsReturn).toBeUndefined();
```
Comment (dverbovyi) [inline src/app/shared/components/product-item/product-item.component.spec.ts]: same as above

File: src/app/shared/components/product-item/product-item.component.spec.ts [hunk 17]
```diff
@@ -25,4 +62,71 @@ describe('ProductItemComponent', () => {
...
+  });
+
+  it('should return undefined if we call stop', () => {
+    fixture = TestBed.createComponent(ProductItemComponent);
+    component = fixture.componentInstance;
+    component.product = product;
+    const functionsReturn = component.onMouseLeaveColor();
+    expect(functionsReturn).toBeUndefined();
```
Comment (dverbovyi) [inline src/app/shared/components/product-item/product-item.component.spec.ts]: same as above

**Review by dverbovyi — CHANGES_REQUESTED** [review]

File: src/app/home-page/components/home-page/home-page.component.scss [hunk 18]
```diff
@@ -3,6 +3,14 @@
 @import '../../../../scss/reset';
 @import '../../../../scss/mixins';
 
+app-product-item-short {
+  width: 30%;
+  margin-right: 5%;
+  &:last-child {
```
Comment (dverbovyi) [inline src/app/home-page/components/home-page/home-page.component.scss:10]: add space above

File: src/app/product-details-page/components/product-description/product-image-preview/zoom/zoom.component.html [hunk 19]
```diff
@@ -0,0 +1,7 @@
+<img src="{{ zsrc }}" [ngClass]="{'hovered': isHovered}" (mousemove)="onMouseMove($event)" (mouseover)="onMouseOver()" (mouseleave)="onMouseLeave()"/>
+<div *ngIf="isHovered"
+  class="zoombox"
+  [style.background-image]="'url(' + zsrc + ')'"
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-description/product-image-preview/zoom/zoom.component.html:4]: use angular string interpolation

"'url({{zsrc}})'"

File: src/app/product-details-page/components/product-description/product-image-preview/zoom/zoom.component.spec.ts [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+
+    const functionsReturn = component.onMouseLeave();
+    expect(functionsReturn).toBeUndefined();
+  });
+
+  it('should return undefined if we call stop', () => {
+    fixture = TestBed.createComponent(ZoomComponent);
+    component = fixture.componentInstance;
+    const functionsReturn = component.onMouseOver();
+    expect(functionsReturn).toBeUndefined();
```
Comment (dverbovyi) [inline src/app/product-details-page/components/product-description/product-image-preview/zoom/zoom.component.spec.ts]: bad test. It validates nothing

**Review by dverbovyi — CHANGES_REQUESTED** [review]

**Review by dverbovyi — APPROVED** [review]
@dnistrova good job!
